src/browser/serial.js

```diff
@@ -42,28 +44,57 @@ function SerialAdapter(element, bus)
     };
     this.init();
 
-
     this.show_char = function(chr)
     {
-        if(chr === "\x08")
-        {
-            this.text = this.text.slice(0, -1);
-            this.update();
-        }
-        else if(chr === "\r")
-        {
-            // do nothing
+        if(this.control_mode === false) {
+            if(chr === "\x08")
+            {
+                this.text = this.text.slice(0, -1);
+                this.update();
+            }
+            else if(chr === "\x1b") {
+               this.control_mode = true;
+               this.control_buffer = "";
+            }
+            else if(chr === "\r")
+            {
+                // do nothing
+            }
+            else
+            {
+               this.text += chr;
+
+               if(chr === "\n")
+               {
+                  this.text_new_line = true;
+               }
+
+               this.update();
+            }
         }
         else
         {
-            this.text += chr;
-
-            if(chr === "\n")
+            this.control_buffer += chr;
+            var cmds = {
+                bksp: /$\[.J^/,
+                clr: /$\[.H\x1B\[.J^/,
+                invalid: /$*[^HJ]^/
```

Comment:
These don't look quite right:

- `$` matches the end of the text and `^` the beginning
- The `*` in `invalid` probably requires something before it
- I believe  `invalid` would match before the other two; it needs to be more specific
- Instead of the `.`, we should probably use a more specific match. [This page](https://www.gnu.org/software/screen/manual/html_node/Control-Sequences.html) says it can be 1, 2 or 3, so `[123]` might be more appropriate

Reply:
Whoops! I'm still getting the hang of regex. Silly mistakes.